Post about Remilia (#465)
crystal-lang/crystal-website: _posts/2023-06-28-Remilia-interview.md

---
title: "Interview wth contributor Remilia Scarlet"
author: beta-zilliani,Remilia

> summary: "<b>Music playback projects in Crystal, games engines, and a bit of Lisp"

Reviewer: ```suggestion
summary: "Music playback projects in Crystal, games engines, and a bit of Lisp"

```

---
<https://remilia.sdf.org/>

> <https://www.youtube.com/watch?v=UX2in-whUik>

Reviewer: Is there a specific reason to put these on top? I would usually expect further reading links at the bottom.

_Conversation between Beta Ziliani and Remilia Scarlet_

> ## A little about you. Who's Remilia?

Reviewer: Interview questions as headlines feel weird. Especially for longer ones that span multiple lines. A better approach would be to use normal paragraphs set in bold. That's the typical format for print interviews.

Reviewer: The "fix it later" approach doesn't work as well as in source repositories. This content is going to be published immediately after merging and we should make sure to polish it right away.
What we do about the old post is a different story.

Reviewer 2: In the current setting adding one level of nesting (`###`) does the trick pretty well:
<img width="945" alt="image" src="https://github.com/crystal-lang/crystal-website/assets/5183634/40506b8b-acad-416f-8baf-46e5282b6523">

And provides the vertical space to separate Q from A. WDYT?

Reviewer: I don't see how this would be any better than the bold paragraph I suggested.

> Well, I'm an open-source developer who has worked with quite a few languages over the years. I mainly do command line tools, but recently I've been interested in audio programming as well.

Reviewer: Unless there's some very specific reason for it, I would suggest to drop such interjections. There's no need for filled pauses in written speach.
```suggestion
I'm an open-source developer who has worked with quite a few languages over the years. I mainly do command line tools, but recently I've been interested in audio programming as well.
```

## You are a Lisper and a Crystalist. That's a strange pairing of languages! What got you into Lisp, and what got you into Crystal?

My interest in Common Lisp started in the mid 2000s just out of curiosity. I kept reading about how great the language could be, and felt that the programs I was working on at the time could benefit from it. So, I sat down, decided "I'm not stopping until I can write a non-trivial program in Lisp", and just went from there. Eventually it became one of my favorite languages. With Crystal, a few years ago I was looking for a language that could provide me with the performance I needed, but that could produce smaller binaries, and that could be easier for others should they want to contribute to my projects (I was the only one I knew who worked with Lisp at the time). After some Wikipedia searching, I discovered Crystal and found that it fit my needs perfectly.
These days, I'd say they're both tied for my favorite language.
## Sometimes you first do a prototype in Lisp, and then a rewrite in Crystal, is that right?

> Yes, that's correct.

Reviewer: Oh that's very shortspoken. Would've been nice to get some more out of this.

Reviewer 3: I can expand on this if you'd like.

Reviewer 2: @MistressRemilia That would be nice. If you want to do it, feel free to paste it here or add it as a suggestion. Then, after that, we'll 🚀 it.

Reviewer 3: Ok, new text is below the line.

----

Yes, that's correct.  Part of this is just due to my history of using Common Lisp.  I've gotten used to the immediacy of it, especially when combined with an Emacs mode called Slime.  You don't just have a REPL, but a live coding environment with the full compiler and image-based development.  I can just put my cursor in a function, do `C-c`, and it'll recompile a function to native code, even while a program is running.  So it makes the adjust->compile->test pipeline much shorter.  Common Lisp is close enough to Crystal with respect to its capabilities and paradigms that moving between them also isn't too big of a deal for me.

However, my Lisp code stays as a prototype in most cases because my Lisp code is never as memory efficient as Crystal.  In most cases I see a roughly 50-75% decrease in memory usage going from my initial Lisp prototype to Crystal.  Shipping a source repo that someone can download and compile also isn't as clean as with Shards, and even if it was, I don't think most users wanting to compile their own software know the idiosyncrasies of the various Lisp compilers out there.  So really, Crystal ends up making a lot more sense to me past the prototype stage in most cases.

Also, a Hello World binary built with Common Lisp is somewhere around 42mb (and can't be run through `strip`), whereas a Hello World in Crystal is only like 300k.

## Tell us about three projects that you're most proud of

> Well, I think the three I'm most proud of right now are [Benben](https://chiselapp.com/user/MistressRemilia/repository/benben/index), a player for a music format called VGM; [midi123](https://chiselapp.com/user/MistressRemilia/repository/midi123/index), a command line MIDI player that can use SoundFonts for audio synthesis; and [RemiAudio](https://chiselapp.com/user/MistressRemilia/repository/remiaudio/index), a general purpose audio processing library that is shared between them.

Reviewer: Ditto

```suggestion
I think the three I'm most proud of right now are [Benben](https://chiselapp.com/user/MistressRemilia/repository/benben/index), a player for a music format called VGM; [midi123](https://chiselapp.com/user/MistressRemilia/repository/midi123/index), a command line MIDI player that can use SoundFonts for audio synthesis; and [RemiAudio](https://chiselapp.com/user/MistressRemilia/repository/remiaudio/index), a general purpose audio processing library that is shared between them.
```

**Completely unrelated, and on a personal note: why Remilia Scarlet?**

Oh, the name? I am a huge fan of a game series called Touhou Project. One of my favorite characters from it is a vampire lady named Remilia. I started going by that online years ago, and got so used to it that I've just adopted it as my actual name. I guess I kind of identified with her lol. My actual real name is Alexa. Going by Remilia also means someone calling out to me won't trigger their Amazon Alexa device.

**Back to your projects, do you know if there are people using them?**

I'm not aware of anyone using my more recent audio-related projects. I'm hoping that will change, given that the libraries behind them could be a good option for music playback in video games.

**You mean, for video games made in Crystal?**

Yes, exactly! I've made a few prototype engines in Crystal, as well as an unfinished port of Doom, and feel that it could be a great language for writing video games based on my experiences with those.

> **About your Doom port! I remember you did some custom-made levels there, right?**

Reviewer 2: This got changed but now it's weird. Changing it back (but I take other suggestions)

```suggestion
**I remember your Doom port! You did some custom-made levels there, right?**
```

Yeah! I've been doing levels for Doom since the mid 90s, and writing my own Doom port has been on my bucket list since I was a kid.

**What's the prospect for the future?**

Benben is likely going to be my main project for the near future. Getting it to support most, or all, of the chips that the VGM format supports is a key goal for the project. I'd also like to backport some of its frontend features into midi123 since the frontends are very similar internally. Aside from that, I'm looking at doing an actual retro-style game, possibly either an FPS or a vertical shooter.

> **The video game story for Crystal is much unexplored. I guess that once Windows support is there, that line will boost. Or at least, my experience is that gamers use Windows…**

Reviewer 2: We can help boost visibility of Windows support:
```suggestion
**The video game story for Crystal is much unexplored. I guess that now that [Windows support is almost there](https://crystal-lang.org/2023/07/06/windows-support-1.9/), that line will boost. Or at least, my experience is that gamers use Windows…**
```
